fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/config/jpa/JpaExceptionHandler.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.core.config.jpa;
+
+import static org.apache.fineract.commands.domain.CommandProcessingResultType.UNDER_PROCESSING;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import javax.servlet.http.HttpServletRequest;
+import org.apache.fineract.infrastructure.core.exception.DuplicateCommandException;
+import org.eclipse.persistence.exceptions.DatabaseException;
+import org.eclipse.persistence.exceptions.ExceptionHandler;
+import org.springframework.beans.BeansException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.request.RequestAttributes;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+@Component
+public class JpaExceptionHandler implements ExceptionHandler, ApplicationContextAware {
```

Comment:
I don't get why you need this class especially with mixing Spring's and JPA's lifecycle.
Cant you just put a try catch around the command execution and unwrap the proper exception there?

Reply:
- in case 2 threads are calling with the same IDK and trying to save the same request initially then a java.sql.SQLIntegrityConstraintViolationException is thrown down the bottom, wrapped in a DatabaseException which is handled by jpa and can not be catched outside -> simply a 500 error will be generated and a dummy result will be sent to the caller, we can not execute custom logic after/before/anywhere of it to save FAILED status or to creat customresult for the caller
- spring is mixed in because Idempotency header's name is configurable therefore we need to access the environment

Reply:
Can you pls share the other implementation where you tried to catch the exception? I'm having doubts to understand why this exception cannot be caught simply on a higher level and handled properly.

On the other hand, if this class is really needed, it still shouldn't be a Bean as well as instantiated by EclipseLink. Mixing objects like that is a smell.
If you need access to the AppContext from a static context, rather create a separate class that's ONLY a Spring Bean and provides access statically.